projects/openapi-cli/src/shapes/closeness/schema-inventory.ts

```diff
@@ -0,0 +1,273 @@
+      }
+    });
+
+    if (closestMatch && closestMatch.percent >= this.closeness) {
+      return closestMatch;
+    } else {
+      return undefined;
+    }
+  }
+
+  async *refsForAdditions(
+    addedPaths: Set<string>,
+    spec: OpenAPIV3.Document
+  ): SpecPatches {
+    if (addedPaths.size === 0) return [];
+    const sorted = Array.from(addedPaths).sort();
+
+    for await (let added of sorted) {
+      const addedSchema = jsonPointerHelpers.get(spec, added);
+      const arrayItems = arrayItemPaths(addedSchema, added);
+
+      const isARootSchema =
+        jsonPointerHelpers.matches(added, [
+          'paths',
+          '**',
+          '**',
+          'responses',
+          '**',
+          'content',
+          '**',
+          'schema',
+        ]) ||
+        // is a root request schema
+        jsonPointerHelpers.matches(added, [
+          'paths',
+          '**',
+          '**',
+          'requestBody',
+          'content',
+          '**',
+          'schema',
+        ]);
+      let matchedRoot = false;
+      let matchedSub = false;
+
+      if (isARootSchema) {
+        const match = this.findClosest(addedSchema);
+        // use ref
+        if (match) {
+          matchedRoot = true;
+          const patch: SpecPatch = {
+            description: `use $ref ${match.ref}`,
+            path: added,
+            impact: [PatchImpact.Refactor],
+            diff: undefined,
+            groupedOperations: [
+              {
+                intent: `use ref at ${added}`,
+                operations: [
+                  {
+                    op: 'replace',
+                    path: added,
+                    value: {
+                      $ref: `#${match.ref}`,
+                    },
+                  },
+                ],
+              },
+            ],
+          };
+          yield patch;
+        }
+      }
+
+      if (!matchedRoot) {
+        for await (let items of arrayItems) {
+          const match = this.findClosest(jsonPointerHelpers.get(spec, items));
+          if (match && match.percent > this.closeness) {
+            matchedSub = true;
+            const patch: SpecPatch = {
+              description: `use $ref ${match.ref}`,
+              path: items,
+              impact: [PatchImpact.Refactor],
+              diff: undefined,
+              groupedOperations: [
+                {
+                  intent: `use ref at ${items}`,
+                  operations: [
+                    {
+                      op: 'replace',
+                      path: items,
+                      value: {
+                        $ref: `#${match.ref}`,
+                      },
+                    },
+                  ],
+                },
+              ],
+            };
+            yield patch;
+          }
+        }
+      }
+      //create a root schema
+      if (!matchedSub && !matchedRoot && isARootSchema) {
+        const refName = refNameGenerator(added);
```

Comment:
Is this always correct to generate ref for a root schema? Perhaps we should only scope this to root schemas that are an object?